libyul/optimiser/StackToMemoryMover.cpp

```diff
@@ -0,0 +1,194 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <libyul/optimiser/StackToMemoryMover.h>
+#include <libyul/optimiser/NameDispenser.h>
+
+#include <libyul/AsmData.h>
+
+#include <libsolutil/CommonData.h>
+#include <libsolutil/Visitor.h>
+
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+namespace {
+void appendMemoryStores(
+	vector<Statement>& _statements,
+	langutil::SourceLocation _loc,
+	vector<pair<YulString, YulString>> const& _mstores
+)
+{
+	for (auto const& [mpos, value]: _mstores)
+		_statements.emplace_back(ExpressionStatement{_loc, FunctionCall{
+			_loc,
+			Identifier{_loc, "mstore"_yulstring},
```

Comment:
Please either assert the dialect or retrieve the name of the mstore function from the dialect.

Reply:
This might have been the reason why I had the assertion in MemoryEscalator left from before splitting this up :-).
I'm wondering about types in this regard... In general properly typed variables will make this whole thing a bit more complex... but we can assume that every dialect has an ``mstore`` that can store the largest type there is in that dialect? Then it'd probably be fine (and once things are properly typed this would just need additional casts... or a slot/offset distinction for smaller types, but that might be crazy :-)).

Reply:
I'll just fix everything to EVMDialect with object access for now.

Reply:
Yes ,this is only for EVM and even if we use a typed dialect, we will strip the types toward the end of the optimization process.